### PR TITLE
New loader policy that features project.ini

### LINKED ISSUE #1250: Associate files and projects
TL;DR: In the specification, what are the rules to: 1. detect that a directory is the root directory of a project? 2. find all the modules of a project.

Currently, file are associated to groups and projects with various heuristics.

* if a file has the same base-name that the directory, then the directory is a group: all nit file in the directory belong to the group
* if a directory and its direct sub-directory are groups, then they belong to the same projects and a group in nested in the other
* if the parent directory of a group is not a group, then the group is the root of a project
* if a directory has a `src` sub-directory then it is a group
* a directory and a `src` sub-directory correspond to the same group

These rules are just crazy and historical and a better scheme is needed, that is KISS and POLA.

A proposal could be to add a specific metadata file in a directory to say that the directory is a project root directory.
So a module can find to which project if belong to

questions:

* how to name it? 
* what kind of metadata? version, author, copyright, compilation options?
* what format for the metadata? xml, json, ini, specific syntax, etc.?

A second need is for a project to find all its modules.
This is used for some commands that can take a project in order to process all its module (eg nitunit)
This will also be used to import modules from the project without having to thing about namespace and directories

Because of this, do we want that any cruft accessible from the project directory is considered a module of the project? Since when I program, I tend to let a lot of broken test-files in directories, my answer was no. There is also the case of generated files. Do we want that all the generated files `foo/.nitunit/test*.nit` belong to the project foo?

Currently, only subdirectory that are themselves groups belong to the same project, therefore this basically limit the project ownership to subdir that have an default module (even empty).
This is the explanation to the various empty `example.nit` files in `lib`. They make the `example` directory qualify as group and this belong to the same project that their parent directory.

For this part of the issue, the right specification seems to be something among:

* accept any source-file accessible from the project root directory
* like the previous but add optional include+exclude rules in the project metadata file
* use a specific group-metadata file to distinguish random-directory from groups and accept in the project only sub directories that are groups (and their subdirectories that are groups)

So, what do people think is the best trade-of as a developer (that use vim nevertheless)?

### PR DESCRIPTION
**Pull request by privat**
This introduce the new loader policy by adapting the existing loader.
The API it still as bad as before but the behavior is expected to be better. Close #1250

The changes impact two decisions made by the loader:

* how to find the project a module belongs to? ie get the root project of an arbitrary file.
* how to find a named module in a project? ie get (and load) the module named `foo` in a mproject object.

The new loader use a recursive directory search for both answers, whereas the older loader only looked at some specific paths.
By the way, the new code use `String::files` and `String::file_stat` (they where unneeded by the previous loader, life was simpler back then). The light FFI used by the bootstrap make this possible.

Here is a detail of the new policy:

* A project is a directory identified with a `project.ini` file.
  Only the existence of the `.ini` file matter, the content is not used yet.
* A group is just a sub-directory of a project; there is no more naming or structural convention (ie hack).
* A module either belongs to a project, or is a standalone module (also called singleton project). This one is not new.

To know the project a module belongs to, its directory is recursively searched up to the `/` directory.
If a project is found, then the module belongs to it.
If no project is found, then the module is a standalone module.

Note: to simplify the migration from the previous loader, a special file `projects.ini` (with a plural) in a directory indicates that sub-directories are projects, even without a `project.ini` file.
These files simplify the PR as only some `projects.ini` files are needed (eg `lib/projects.ini`)

To search for a module in a project, the named module is searched in the directory and subdirectories of the project.

This PR also expose some vocabulary for the end user. Maybe some terms could be improved before the PR is merged. Some propositions:

* rename `project` as `package` (so `package.ini` instead of `project.ini`). Might be non-POLA from people coming from a Java ecosystem.
* rename `group` as `directory`. With the new policy, groups are more transparent than ever and are only here for the programmer to group modules as he prefers.

### CODE REVIEW

**Comment by privat** [comment]
reroll: fix an issue with the android platform that assumed that the root group of a project was always `src`

**Comment by R4PaSs** [comment]
:shipit:

File: src/loader.nit [hunk 1]
```diff
@@ -438,42 +388,62 @@ redef class ModelBuilder
...
+		if ini == null then
+			# No ini, multiple course of action
+
+			# The root of the directory hierarchy in the file system.
+			if rdp == "/" then
+				mgroups[rdp] = null
+				return null
+			end
+
+			# Special stopper `projects.nit`
```
Comment (xymus) [inline src/loader.nit]: .nit or .ini?

**Comment by xymus** [comment]
:shipit:

**Comment by Morriar** [comment]
Is the cover letter of this PR the only place where we can find documentation about the "`projects.ini` stopper"?

File: src/loader.nit [hunk 2]
```diff
@@ -966,6 +936,15 @@ class ModulePath
 	redef fun to_s do return filepath
 end
 
+redef class MProject
+	# The associated `.ini` file, if any
+	#
+	# The `ini` file is presented and might contains invalid or missing information.
```
Comment (Morriar) [inline src/loader.nit]: contain*

Comment (Morriar) [inline src/loader.nit]: Reading this a second time, I don't understand the `is presented`. What are you trying to say here?

**Comment by privat** [comment]
reroll: typos and improve doc

@Morriar the `projects.ini` is documented in the cover-letter and in the code. It is a temporary feature to ease the migration of existing code.
Other place to document it are the manual (but #761 is old) and NEPs, unfortunately, the loaded was never NEPified (spec are in #1250 and this PR).

**Pull request merged**